### PR TITLE
FE - e2e - Cover dashboard filters with tests

### LINKED ISSUE #46527: FE - e2e - Cover dashboard filters with tests

### LINKED ISSUE #47304: Unit of time parameter does not work when used together with another filter
CI run with failure: https://github.com/metabase/metabase/actions/runs/10578852932/job/29310334408?pr=46958

1. `git checkout 46527-dashboard-filters-tests`
2. Run e2e test called `should not overwrite parameter mappings for a card when doing auto-wiring`

The test will fail because unit of time parameter won't be applied by BE when calling POST `/api/dashboard/:id/dashcard/:id/card/:id/query`

----

I think the discrepancy comes from the fact that FE does not call `Lib.ensureFilterStage` for unit of time parameters (see [code](https://github.com/metabase/metabase/blob/37d993b4e555ec0beff2185d9d67050171f79fbf/frontend/src/metabase-lib/v1/parameters/utils/targets.ts#L172-L182)). The reason is that unit of time is a parameter, not a filter (i.e. it does not produce a filter clause). We don't want to change how unit of parameters work as a part of this epic, although enabling passing unit of filter parameter at other query stages than the last one has been discussed ([see bottom of product doc](https://www.notion.so/metabase/Allow-to-use-both-pre-and-post-last-aggregation-columns-in-dashboard-filters-b32a6bb5cebc4723b8b9b81870217a79?pvs=4#38f1f3f47d98427eaa8994ad7cfc8289)).

I already adjusted FE to not send `stage-number` option with unit of time parameters (37d993b4e555ec0beff2185d9d67050171f79fbf).
Now BE needs to figure out which stage to apply the unit of time parameter to.

### PR DESCRIPTION
**Pull request by kamilmielnik**
Closes #46527
Closes #47304

Part of testing plan: #46520
Epic: #46519

### Description

This PR adds e2e tests that assert which columns dashboard filters can be mapped to. It tests the following dimensions:
- entity type
- number of query stages
- presence of aggregations and/or breakouts in last query stage

This test suite will also serve as a boilerplate for the upcoming tasks (see section below).
It will grow and queries will be reused.

BE changes in this PR are a fix for #47304

#### What's remaining according to the testing plan

- #47219 - applying the filters in dashboard and title drills for Q1-Q8 for every column source
- #46528 - custom click behavior for Q1-Q8 for every column source
- #47234

### How to verify

CI is green except 1 failing test due to #47304.

### CODE REVIEW

File: e2e/test/scenarios/dashboard-filters/dashboard-filters-query-stages.cy.spec.ts [hunk 1]
```diff
@@ -0,0 +1,798 @@
...
+    }
+  });
+
+  describe("1-stage queries", () => {
+    describe("Q1 - join, custom column, no aggregations, no breakouts", () => {
+      beforeEach(() => {
+        createAndVisitDashboardWithQueryMatrix(createQ1Query);
+      });
+
+      it("allows to map to all relevant columns", () => {
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard-filters/dashboard-filters-query-stages.cy.spec.ts:206]: The plan:
- In #47219 I will test applying the filters in dashboard and title drills for Q1-Q8 (separate test for every column source possible)
- In #46528 I will test custom click behavior for Q1-Q8 for every column source possible

File: e2e/test/scenarios/dashboard-filters/dashboard-filters-query-stages.cy.spec.ts [hunk 2]
```diff
@@ -0,0 +1,916 @@
...
+      verifyDashcardMappingOptions(BASE_MODEL_INDEX, [
+        ["Base Orders Model", ORDERS_NUMBER_COLUMNS],
+        ["Product", PRODUCTS_NUMBER_COLUMNS],
+        ["User", PEOPLE_NUMBER_COLUMNS],
+      ]);
+    }
+  });
+
+  describe("1-stage queries", () => {
+    describe("Q1 - join, custom column, no aggregations, no breakouts", () => {
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard-filters/dashboard-filters-query-stages.cy.spec.ts:202]: Every Q (Query) is tested against the matrix of 4 cases:
- card is a question / card is a model
- source table is a question / source table is a model

File: e2e/test/scenarios/dashboard-filters/dashboard-filters-query-stages.cy.spec.ts [hunk 3]
```diff
@@ -0,0 +1,1298 @@
...
+}
+
+// Q1 - join, custom column, no aggregations, no breakouts
+function createQ1Query(source: Card): StructuredQuery {
+  return {
+    "source-table": `card__${source.id}`,
+    expressions: {
+      Net: ["-", TOTAL_FIELD, TAX_FIELD],
+    },
+    joins: [
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard-filters/dashboard-filters-query-stages.cy.spec.ts:990]: We never use `metabase-lib` in e2e tests to create queries but maybe we should start :thinking:

File: src/metabase/query_processor/middleware/parameters/mbql.clj [hunk 4]
```diff
@@ -94,12 +94,15 @@
...
+   {[_dimension [_field target-field-id {:keys [base-type temporal-unit]}] dim-opts] :target
     :keys [value] :as _param}]
   (let [new-unit (keyword value)
         base-type (or base-type
                       (when (integer? target-field-id)
-                        (:base-type (lib.metadata/field (qp.store/metadata-provider) target-field-id))))]
+                        (:base-type (lib.metadata/field (qp.store/metadata-provider) target-field-id))))
+        stage-path (if-let [stage-number (:stage-number dim-opts)]
+                     (mbql.u/stage-path (:query query) stage-number)
+                     [])]
```
Comment (bshepherdson) [inline src/metabase/query_processor/middleware/parameters/mbql.clj:105]: Perhaps `mbql.u/stage-path` itself should respond to a stage number of `nil` with `[]`, ie. the last stage/outermost query?

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/query_processor/middleware/parameters/mbql.clj (hunk 4 above)
Comment (metamben) [reply src/metabase/query_processor/middleware/parameters/mbql.clj:105]: That's a good idea, thanks!

**Review by romeovs — APPROVED** [review]

**Pull request merged**